insights/client/crypto.py

```diff
@@ -0,0 +1,122 @@
+import os.path
+import logging
+import shutil
+import tempfile
+import shlex
+import subprocess
+
+from .constants import InsightsConstants as constants
+
+logger = logging.getLogger(__file__)
+
+
+def _call_gpg(home, command):
+    """
+    Run /usr/bin/gpg.
+
+    Parameters:
+        home (str): Path to GPG's home directory.
+        command (str): Actual arguments passed to the GPG binary.
+
+    Returns (str, str, int):
+        Standard output, standard error and the return code.
+    """
+    command = (
+        "/usr/bin/gpg --homedir {home} {command}"
+    ).format(home=home, command=command)
+    process = subprocess.Popen(shlex.split(command), stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    raw_stdout, raw_stderr = process.communicate()
+    stdout, stderr = raw_stdout.decode("utf-8"), raw_stderr.decode("utf-8")
+
+    logger.debug(command)
+    logger.debug("status code {code}".format(code=process.returncode))
+
+    if process.returncode > 0 and stdout:
+        logger.debug("stdout is not empty:\n%s", stdout.strip())
+    if process.returncode > 0 and stderr:
+        logger.debug("stderr is not empty:\n%s", stderr.strip())
+
+    return stdout, stderr, process.returncode
+
+
+def verify_gpg_signed_file(file, signature):
+    """
+    Verify a file that was signed using GPG.
+
+    Parameters:
+        file (str): A path to the signed file.
+        signature (str): A path to the detached signature.
+
+    Returns (dict):
+        ok (str): Whether the signature holds.
+        stdout (str): Standard output of GPG.
+        stderr (str): Standard error of GPG.
+        code (int): Return code of GPG.
+    """
+    if not os.path.isfile(file):
+        message = "cannot verify signature of '{file}', file does not exist".format(file=file)
+        logger.debug(message)
+        return {
+            'gpg': False,
+            'stdout': '',
+            'stderr': '',
+            'rc': 1,
+        }
+
+    if not os.path.isfile(signature):
+        message = (
+            "cannot verify signature of '{file}', signature '{signature}' does not exist"
+        ).format(file=file, signature=signature)
+        logger.debug(message)
+        return {
+            'gpg': False,
+            'stdout': '',
+            'stderr': message,
+            'rc': 1,
+        }
+
+    keys = [constants.pub_gpg_path]
+    _extra_key = os.getenv("EXTRA_GPG_KEY")
+    if _extra_key:
+        keys.append(_extra_key)
+
+    for key in keys:
+        if not os.path.isfile(key):
+            message = (
+                "cannot verify signature of '{file}', key '{key}' does not exist"
+            ).format(file=file, key=key)
+            logger.debug(message)
+            return {
+                'gpg': False,
+                'stdout': '',
+                'stderr': message,
+                'rc': 1,
+            }
+
+    home = tempfile.mkdtemp()
```

Comment:
If the minimum version of Python is high enough, you can use `NamedTemporaryDirectory` as a context manager to automatically delete this directory: https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory

If not, you should probably use a `try:` `finally:` with this directory and call `shutil.rmtree()` from the finally block.  Note that I've had issues with gpg creating its own temporary files in its home directory and then `shutil.rmtree()` ends up racing with gpg's own deletion of those files (resulting in `shutil.rmtree()` erroring out because the file it was going to delete has already been deleted by gpg) so you may need to protect against that here as well.

Reply:
Pino says that this code has to be compatible with python-2.6+ so the `try:` `finally:` is the way to go here.  

For reference, here's my proposal for a similar `try:` `finally:` in convert2rhel code that tries to deal with the race condition I mentioned: https://github.com/oamg/convert2rhel/pull/922/files#diff-ed1ca2e4a8b27f1bee13d608562d4e94eceb2972790eccf4d84f366b750f221d  (Note: It uses a `for-else` in addition to the more common `except-else`)